File: lib/src/contentId.test.ts
import { describe, expect, it } from "vitest";
import {
  byIdPrefix,
  formatContentId,
  masterPlaylistKey,
  metadataKey,
  parseContentId,
} from "./contentId.js";

describe("formatContentId / parseContentId", () => {
  it("round-trips sha256", () => {
    const id = formatContentId("sha256", "abc123");
    expect(id).toBe("sha256:abc123");
    expect(parseContentId(id)).toEqual({ scheme: "sha256", id: "abc123" });
  });

  it("round-trips psig", () => {
    expect(parseContentId("psig:xyz")).toEqual({ scheme: "psig", id: "xyz" });
  });

  it("rejects missing scheme prefix", () => {
    expect(() => parseContentId("abc123")).toThrow(/missing scheme/i);
  });

  it("rejects unknown scheme", () => {
    expect(() => parseContentId("md5:abc")).toThrow(/Unknown content ID scheme/);
  });
});

describe("canonical paths", () => {
  const id = "sha256:abc";
  it("byIdPrefix", () => {
    expect(byIdPrefix(id)).toBe("by-id/sha256:abc/");
  });
  it("masterPlaylistKey", () => {
    expect(masterPlaylistKey(id)).toBe("by-id/sha256:abc/master.m3u8");
  });
  it("metadataKey", () => {
    expect(metadataKey(id)).toBe("by-id/sha256:abc/metadata.json");
  });
});

File: core/contentid_test.go
package core

import (
	"strings"
	"testing"
)

func TestFormatParseContentID(t *testing.T) {
	id := FormatContentID(SchemeSHA256, "abc123")
	if id != "sha256:abc123" {
		t.Fatalf("FormatContentID = %q", id)
	}
	scheme, body, err := ParseContentID(id)
	if err != nil || scheme != SchemeSHA256 || body != "abc123" {
		t.Fatalf("ParseContentID = (%q,%q,%v)", scheme, body, err)
	}
	if s, b, err := ParseContentID("psig:xyz"); err != nil || s != SchemePSig || b != "xyz" {
		t.Fatalf("ParseContentID psig = (%q,%q,%v)", s, b, err)
	}
	if _, _, err := ParseContentID("abc123"); err == nil || !strings.Contains(strings.ToLower(err.Error()), "missing scheme") {
		t.Fatalf("expected missing-scheme error, got %v", err)
	}
	if _, _, err := ParseContentID("md5:abc"); err == nil || !strings.Contains(err.Error(), "unknown content ID scheme") {
		t.Fatalf("expected unknown-scheme error, got %v", err)
	}
}

func TestCanonicalPaths(t *testing.T) {
	id := "sha256:abc"
	if got := ByIDPrefix(id); got != "by-id/sha256:abc/" {
		t.Errorf("ByIDPrefix = %q", got)
	}
	if got := MasterPlaylistKey(id); got != "by-id/sha256:abc/master.m3u8" {
		t.Errorf("MasterPlaylistKey = %q", got)
	}
	if got := MetadataKey(id); got != "by-id/sha256:abc/metadata.json" {
		t.Errorf("MetadataKey = %q", got)
	}
}
